drivers/note/noteram_driver.c

```diff
@@ -0,0 +1,477 @@
+
+          if (retlen == 0)
+            {
+              /* If nothing was read then report the error.  Otherwise,
+               * just silently drop the note.
+               */
+
+              retlen = notelen;
+            }
+
+          break;
+        }
+
+      /* Update pointers from the note that was transferred */
+
+      retlen += notelen;
+      buffer += notelen;
+      buflen -= notelen;
+
+      /* Will the next note fit?  There is a race here and even if the next
+       * note will fit, it may fail still when noteram_get() is called.
+       *
+       * It won't fit (or an error occurred).  Return what we have without
+       * trying to get the next note (which would cause it to be deleted).
+       */
+
+      notelen = noteram_size();
+    }
+  while (notelen > 0 && notelen <= buflen);
+
+  sched_unlock();
+  return retlen;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: note_add
+ *
+ * Description:
+ *   Add the variable length note to the transport layer
+ *
+ * Input Parameters:
+ *   note    - The note buffer
+ *   notelen - The buffer length
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *   We are within a critical section.
+ *
+ ****************************************************************************/
+
+void note_add(FAR const uint8_t *note, uint8_t notelen)
+{
+  unsigned int head;
+  unsigned int next;
+
+#ifdef CONFIG_SMP
+  irqstate_t flags = up_irq_save();
+  spin_lock_wo_note(&g_noteram_lock);
+#endif
+
+  /* Get the index to the head of the circular buffer */
+
+  DEBUGASSERT(note != NULL && notelen < CONFIG_DRIVER_NOTERAM_BUFSIZE);
+  head = g_noteram_info.ni_head;
+
+  /* Loop until all bytes have been transferred to the circular buffer */
+
+  while (notelen > 0)
+    {
+      /* Get the next head index.  Would it collide with the current tail
+       * index?
+       */
+
+      next = noteram_next(head, 1);
+      if (next == g_noteram_info.ni_tail)
+        {
+          /* Yes, then remove the note at the tail index */
+
+          noteram_remove();
+        }
+
+      /* Save the next byte at the head index */
+
+      g_noteram_info.ni_buffer[head] = *note++;
+
+      head = next;
+      notelen--;
+    }
+
+  g_noteram_info.ni_head = head;
+
+#ifdef CONFIG_SMP
+  spin_unlock_wo_note(&g_noteram_lock);
+  up_irq_restore(flags);
+#endif
+}
+
+/****************************************************************************
+ * Name: noteram_register
+ *
+ * Description:
+ *   Register a serial driver at /dev/note that can be used by an
+ *   application to read data from the circular note buffer.
+ *
+ * Input Parameters:
+ *   None.
+ *
+ * Returned Value:
+ *   Zero is returned if the circular buffer is empty.  Otherwise, a negated
+ *   errno value is returned.
```

Comment:
This comment is wrong. Zero is returned if the device is registered successfully.

Reply:
Done.